In [ ]:

import pandas as pd
from google.colab import files

uploaded = files.upload()

Saving archive.zip to archive.zip


In [ ]:

import zipfile
import os

with zipfile.ZipFile("archive.zip", "r") as zip_ref:
    zip_ref.extractall("dataset")

print("ZIP file successfully extracted!")
print(os.listdir("dataset"))

ZIP file successfully extracted!
['dirty_cafe_sales.csv']


In [ ]:

import pandas as pd

df = pd.read_csv("dataset/dirty_cafe_sales.csv")

print("Dataset loaded successfully!")
print("Rows and Columns:", df.shape)

df.head()

Dataset loaded successfully!
Rows and Columns: (10000, 8)


,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11


In [ ]:

print("----- DATASET INFORMATION -----")
df.info()

print("\n----- MISSING VALUES -----")
print(df.isnull().sum())

print("\n----- DUPLICATE ROWS -----")
print(df.duplicated().sum())

print("\n----- BASIC STATISTICS -----")
df.describe(include="all")

----- DATASET INFORMATION -----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price Per Unit    9821 non-null   object
 4   Total Spent       9827 non-null   object
 5   Payment Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB

----- MISSING VALUES -----
Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

----- DUPLICATE ROWS -----
0

----- BASIC STATISTICS -----


,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
count,10000,9667,9862,9821,9827,7421,6735,9841
unique,10000,10,7,8,19,5,4,367
top,TXN_9226047,Juice,5,3.0,6.0,Digital Wallet,Takeaway,UNKNOWN
freq,1,1171,2013,2429,979,2291,3022,159


In [ ]:

print("----- UNIQUE VALUES -----")

for column in df.columns:
    print(f"\n{column}:")
    print(df[column].dropna().unique()[:15])

----- UNIQUE VALUES -----

Transaction ID:
['TXN_1961373' 'TXN_4977031' 'TXN_4271903' 'TXN_7034554' 'TXN_3160411'
 'TXN_2602893' 'TXN_4433211' 'TXN_6699534' 'TXN_4717867' 'TXN_2064365'
 'TXN_2548360' 'TXN_3051279' 'TXN_7619095' 'TXN_9437049' 'TXN_8915701']

Item:
['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' 'UNKNOWN' 'Sandwich' 'ERROR'
 'Juice' 'Tea']

Quantity:
['2' '4' '5' '3' '1' 'ERROR' 'UNKNOWN']

Price Per Unit:
['2.0' '3.0' '1.0' '5.0' '4.0' '1.5' 'ERROR' 'UNKNOWN']

Total Spent:
['4.0' '12.0' 'ERROR' '10.0' '20.0' '9.0' '16.0' '15.0' '25.0' '8.0' '5.0'
 '3.0' '6.0' 'UNKNOWN' '2.0']

Payment Method:
['Credit Card' 'Cash' 'UNKNOWN' 'Digital Wallet' 'ERROR']

Location:
['Takeaway' 'In-store' 'UNKNOWN' 'ERROR']

Transaction Date:
['2023-09-08' '2023-05-16' '2023-07-19' '2023-04-27' '2023-06-11'
 '2023-03-31' '2023-10-06' '2023-10-28' '2023-07-28' '2023-12-31'
 '2023-11-07' 'ERROR' '2023-05-03' '2023-06-01' '2023-03-21']


In [ ]:

print("----- ERROR AND UNKNOWN COUNT -----")

for column in df.columns:
    error_count = (df[column] == "ERROR").sum()
    unknown_count = (df[column] == "UNKNOWN").sum()

    print(f"\n{column}")
    print("ERROR:", error_count)
    print("UNKNOWN:", unknown_count)

----- ERROR AND UNKNOWN COUNT -----

Transaction ID
ERROR: 0
UNKNOWN: 0

Item
ERROR: 292
UNKNOWN: 344

Quantity
ERROR: 170
UNKNOWN: 171

Price Per Unit
ERROR: 190
UNKNOWN: 164

Total Spent
ERROR: 164
UNKNOWN: 165

Payment Method
ERROR: 306
UNKNOWN: 293

Location
ERROR: 358
UNKNOWN: 338

Transaction Date
ERROR: 142
UNKNOWN: 159


In [ ]:

import numpy as np

# Create a copy of the original dataset
df_clean = df.copy()

# Replace ERROR and UNKNOWN with missing values
df_clean = df_clean.replace(["ERROR", "UNKNOWN"], np.nan)

print("ERROR and UNKNOWN replaced with missing values!")

# Check missing values after replacement
print("\n----- MISSING VALUES AFTER REPLACEMENT -----")
print(df_clean.isnull().sum())

ERROR and UNKNOWN replaced with missing values!

----- MISSING VALUES AFTER REPLACEMENT -----
Transaction ID         0
Item                 969
Quantity             479
Price Per Unit       533
Total Spent          502
Payment Method      3178
Location            3961
Transaction Date     460
dtype: int64


In [ ]:

# Convert numeric columns to numbers
numeric_columns = ["Quantity", "Price Per Unit", "Total Spent"]

for column in numeric_columns:
    df_clean[column] = pd.to_numeric(
        df_clean[column], errors="coerce"
    )

# Convert date column to datetime
df_clean["Transaction Date"] = pd.to_datetime(
    df_clean["Transaction Date"], errors="coerce"
)

# Convert Transaction ID to string
df_clean["Transaction ID"] = df_clean["Transaction ID"].astype("string")

print("Data types corrected!")

print("\n----- UPDATED DATA TYPES -----")
print(df_clean.dtypes)

Data types corrected!

----- UPDATED DATA TYPES -----
Transaction ID      string[python]
Item                        object
Quantity                   float64
Price Per Unit             float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
dtype: object


In [ ]:

# Calculate missing values and percentages

missing_report = pd.DataFrame({
    "Missing Values": df_clean.isnull().sum(),
    "Missing Percentage": (
        df_clean.isnull().sum() / len(df_clean) * 100
    ).round(2)
})

print("----- MISSING VALUES REPORT -----")
print(missing_report)

----- MISSING VALUES REPORT -----
                  Missing Values  Missing Percentage
Transaction ID                 0                0.00
Item                         969                9.69
Quantity                     479                4.79
Price Per Unit               533                5.33
Total Spent                  502                5.02
Payment Method              3178               31.78
Location                    3961               39.61
Transaction Date             460                4.60


In [ ]:

# Fill missing categorical values

categorical_columns = ["Item", "Payment Method", "Location"]

for column in categorical_columns:
    df_clean[column] = df_clean[column].fillna("Unknown")

print("Categorical missing values filled!")

print(df_clean[categorical_columns].isnull().sum())

Categorical missing values filled!
Item              0
Payment Method    0
Location          0
dtype: int64


In [ ]:
# Fill missing numeric values with median

numeric_fill_columns = ["Quantity", "Price Per Unit"]

for column in numeric_fill_columns:
    median_value = df_clean[column].median()
    df_clean[column] = df_clean[column].fillna(median_value)

print("Quantity and Price Per Unit missing values filled!")

print("\n----- REMAINING MISSING VALUES -----")
print(df_clean.isnull().sum())


Quantity and Price Per Unit missing values filled!

----- REMAINING MISSING VALUES -----
Transaction ID        0
Item                  0
Quantity              0
Price Per Unit        0
Total Spent         502
Payment Method        0
Location              0
Transaction Date    460
dtype: int64


In [ ]:
# Calculate Total Spent using Quantity × Price Per Unit

calculated_total = (
    df_clean["Quantity"] * df_clean["Price Per Unit"]
)

df_clean["Total Spent"] = df_clean["Total Spent"].fillna(
    calculated_total
)

print("Total Spent missing values handled!")

print("\n----- REMAINING MISSING VALUES -----")
print(df_clean.isnull().sum())

Total Spent missing values handled!

----- REMAINING MISSING VALUES -----
Transaction ID        0
Item                  0
Quantity              0
Price Per Unit        0
Total Spent           0
Payment Method        0
Location              0
Transaction Date    460
dtype: int64


In [ ]:
# Check duplicate rows

duplicate_count = df_clean.duplicated().sum()

print("----- DUPLICATE CHECK -----")
print("Duplicate rows:", duplicate_count)

# Check missing values

print("\n----- MISSING VALUES CHECK -----")
print(df_clean.isnull().sum())

# Check data types

print("\n----- DATA TYPES CHECK -----")
print(df_clean.dtypes)

----- DUPLICATE CHECK -----
Duplicate rows: 0

----- MISSING VALUES CHECK -----
Transaction ID        0
Item                  0
Quantity              0
Price Per Unit        0
Total Spent           0
Payment Method        0
Location              0
Transaction Date    460
dtype: int64

----- DATA TYPES CHECK -----
Transaction ID      string[python]
Item                        object
Quantity                   float64
Price Per Unit             float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
dtype: object


In [ ]:
# Outlier check using IQR method

numeric_columns = ["Quantity", "Price Per Unit", "Total Spent"]

print("----- OUTLIER CHECK -----")

for column in numeric_columns:
    Q1 = df_clean[column].quantile(0.25)
    Q3 = df_clean[column].quantile(0.75)

    IQR = Q3 - Q1

    lower_limit = Q1 - 1.5 * IQR
    upper_limit = Q3 + 1.5 * IQR

    outliers = df_clean[
        (df_clean[column] < lower_limit) |
        (df_clean[column] > upper_limit)
    ]

    print(f"\n{column}")
    print("Lower Limit:", lower_limit)
    print("Upper Limit:", upper_limit)
    print("Outliers:", len(outliers))

----- OUTLIER CHECK -----

Quantity
Lower Limit: -1.0
Upper Limit: 7.0
Outliers: 0

Price Per Unit
Lower Limit: -1.0
Upper Limit: 7.0
Outliers: 0

Total Spent
Lower Limit: -8.0
Upper Limit: 24.0
Outliers: 268


In [ ]:
# Inspect Total Spent outliers

Q1 = df_clean["Total Spent"].quantile(0.25)
Q3 = df_clean["Total Spent"].quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR

outliers_df = df_clean[
    (df_clean["Total Spent"] < lower_limit) |
    (df_clean["Total Spent"] > upper_limit)
]

print("Total Spent Outliers:", len(outliers_df))

print("\n----- OUTLIER DATA -----")
display(outliers_df.head(20))

Total Spent Outliers: 268

----- OUTLIER DATA -----


,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
10,TXN_2548360,Salad,5.0,5.0,25.0,Cash,Takeaway,2023-11-07
51,TXN_6342161,Salad,5.0,5.0,25.0,Unknown,Takeaway,2023-01-08
52,TXN_8914892,Unknown,5.0,5.0,25.0,Digital Wallet,Unknown,2023-03-15
96,TXN_5220895,Salad,5.0,5.0,25.0,Cash,In-store,2023-06-10
100,TXN_9517146,Unknown,5.0,5.0,25.0,Cash,Takeaway,2023-10-30
150,TXN_8687151,Salad,5.0,5.0,25.0,Cash,Unknown,2023-06-10
157,TXN_4283157,Salad,5.0,5.0,25.0,Digital Wallet,In-store,2023-11-25
177,TXN_1896955,Salad,3.0,5.0,25.0,Unknown,In-store,2023-09-23
214,TXN_8693704,Salad,3.0,5.0,25.0,Cash,In-store,2023-05-04
330,TXN_5523450,Salad,5.0,5.0,25.0,Credit Card,Takeaway,2023-07-02


In [ ]:
# Check Total Spent consistency

df_clean["Calculated Total"] = (
    df_clean["Quantity"] * df_clean["Price Per Unit"]
)

df_clean["Difference"] = (
    df_clean["Total Spent"] - df_clean["Calculated Total"]
)

inconsistent_rows = df_clean[
    df_clean["Difference"].abs() > 0.01
]

print("----- TOTAL SPENT CONSISTENCY CHECK -----")
print("Inconsistent rows:", len(inconsistent_rows))

display(inconsistent_rows.head(10))

----- TOTAL SPENT CONSISTENCY CHECK -----
Inconsistent rows: 733


,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date,Calculated Total,Difference
20,TXN_3522028,Smoothie,3.0,4.0,20.0,Cash,In-store,2023-04-04,12.0,8.0
55,TXN_5522862,Cookie,3.0,1.0,2.0,Credit Card,Takeaway,2023-03-19,3.0,-1.0
57,TXN_2080895,Cake,3.0,3.0,3.0,Digital Wallet,In-store,2023-04-19,9.0,-6.0
66,TXN_8501819,Juice,3.0,3.0,6.0,Cash,Unknown,2023-03-30,9.0,-3.0
68,TXN_8427104,Salad,2.0,3.0,10.0,Unknown,In-store,2023-10-27,6.0,4.0
85,TXN_8035512,Tea,3.0,3.0,4.5,Cash,Unknown,2023-10-29,9.0,-4.5
147,TXN_9336980,Salad,4.0,3.0,20.0,Cash,In-store,2023-06-06,12.0,8.0
151,TXN_4031509,Unknown,4.0,3.0,16.0,Credit Card,Takeaway,2023-01-04,12.0,4.0
153,TXN_6541415,Unknown,3.0,3.0,12.0,Cash,In-store,2023-11-25,9.0,3.0
162,TXN_9238666,Cookie,1.0,3.0,1.0,Unknown,Takeaway,2023-07-31,3.0,-2.0


In [ ]:
# Correct Total Spent using Quantity × Price Per Unit

df_clean["Total Spent"] = (
    df_clean["Quantity"] * df_clean["Price Per Unit"]
)

print("Total Spent values corrected!")

# Verify consistency

df_clean["Calculated Total"] = (
    df_clean["Quantity"] * df_clean["Price Per Unit"]
)

df_clean["Difference"] = (
    df_clean["Total Spent"] - df_clean["Calculated Total"]
)

inconsistent_rows = df_clean[
    df_clean["Difference"].abs() > 0.01
]

print("Remaining inconsistent rows:", len(inconsistent_rows))

Total Spent values corrected!
Remaining inconsistent rows: 0


In [ ]:
# FINAL DATA QUALITY CHECK

print("----- FINAL QUALITY CHECK -----")

# Missing values
print("\nMissing Values:")
print(df_clean.isnull().sum())

# Duplicate rows
print("\nDuplicate Rows:")
print(df_clean.duplicated().sum())

# Negative values
print("\nNegative Quantity:", (df_clean["Quantity"] < 0).sum())
print("Negative Price:", (df_clean["Price Per Unit"] < 0).sum())
print("Negative Total:", (df_clean["Total Spent"] < 0).sum())

# Dataset shape
print("\nDataset Shape:", df_clean.shape)

# Data types
print("\nData Types:")
print(df_clean.dtypes)

----- FINAL QUALITY CHECK -----

Missing Values:
Transaction ID        0
Item                  0
Quantity              0
Price Per Unit        0
Total Spent           0
Payment Method        0
Location              0
Transaction Date    460
Calculated Total      0
Difference            0
dtype: int64

Duplicate Rows:
0

Negative Quantity: 0
Negative Price: 0
Negative Total: 0

Dataset Shape: (10000, 10)

Data Types:
Transaction ID      string[python]
Item                        object
Quantity                   float64
Price Per Unit             float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
Calculated Total           float64
Difference                 float64
dtype: object


In [ ]:
# FINAL DATA QUALITY CHECK

print("----- FINAL QUALITY CHECK -----")

# Missing values
print("\nMissing Values:")
print(df_clean.isnull().sum())

# Duplicate rows
print("\nDuplicate Rows:")
print(df_clean.duplicated().sum())

# Negative values
print("\nNegative Quantity:", (df_clean["Quantity"] < 0).sum())
print("Negative Price:", (df_clean["Price Per Unit"] < 0).sum())
print("Negative Total:", (df_clean["Total Spent"] < 0).sum())

# Dataset shape
print("\nDataset Shape:", df_clean.shape)

# Data types
print("\nData Types:")
print(df_clean.dtypes)

----- FINAL QUALITY CHECK -----

Missing Values:
Transaction ID        0
Item                  0
Quantity              0
Price Per Unit        0
Total Spent           0
Payment Method        0
Location              0
Transaction Date    460
Calculated Total      0
Difference            0
dtype: int64

Duplicate Rows:
0

Negative Quantity: 0
Negative Price: 0
Negative Total: 0

Dataset Shape: (10000, 10)

Data Types:
Transaction ID      string[python]
Item                        object
Quantity                   float64
Price Per Unit             float64
Total Spent                float64
Payment Method              object
Location                    object
Transaction Date    datetime64[ns]
Calculated Total           float64
Difference                 float64
dtype: object


In [ ]:
# Remove temporary helper columns

df_clean = df_clean.drop(
    columns=["Calculated Total", "Difference"]
)

print("Temporary columns removed!")

print("\nFinal Columns:")
print(df_clean.columns.tolist())

print("\nFinal Dataset Shape:")
print(df_clean.shape)

Temporary columns removed!

Final Columns:
['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent', 'Payment Method', 'Location', 'Transaction Date']

Final Dataset Shape:
(10000, 8)


In [ ]:
# Export cleaned dataset

output_file = "cleaned_cafe_sales.csv"

df_clean.to_csv(output_file, index=False)

print("Cleaned CSV exported successfully!")
print("File name:", output_file)

Cleaned CSV exported successfully!
File name: cleaned_cafe_sales.csv


In [ ]:
# BEFORE VS AFTER CLEANING SUMMARY

summary = pd.DataFrame({
    "Metric": [
        "Total Rows",
        "Total Columns",
        "Duplicate Rows",
        "Missing Values"
    ],
    "Before Cleaning": [
        df.shape[0],
        df.shape[1],
        df.duplicated().sum(),
        df.isnull().sum().sum()
    ],
    "After Cleaning": [
        df_clean.shape[0],
        df_clean.shape[1],
        df_clean.duplicated().sum(),
        df_clean.isnull().sum().sum()
    ]
})

print("----- BEFORE VS AFTER CLEANING -----")
display(summary)

----- BEFORE VS AFTER CLEANING -----


,Metric,Before Cleaning,After Cleaning
0,Total Rows,10000,10000
1,Total Columns,8,8
2,Duplicate Rows,0,0
3,Missing Values,6826,460


## Data Cleaning Observations

1. The dataset contained 10,000 rows and 8 columns.
2. ERROR and UNKNOWN values were converted into missing values.
3. Missing categorical values were replaced with "Unknown".
4. Missing Quantity and Price Per Unit values were filled using the median.
5. Missing and inconsistent Total Spent values were corrected using Quantity × Price Per Unit.
6. No duplicate rows were found.
7. No negative values were found in numeric columns.
8. Outliers in Total Spent were inspected and retained as valid transactions.
9. Missing Transaction Date values (460) were retained as NaT because the original dates were unavailable.
10. The cleaned dataset was exported as cleaned_cafe_sales.csv.